File: typescript/src/exceptions.ts
/**
 * Base exception for all DF SDK errors.
 */
export class DFError extends Error {
  constructor(message: string) {
    super(message);
    this.name = this.constructor.name;
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

/**
 * Raised when the API returns a 401 Unauthorized.
 */
export class AuthenticationError extends DFError {}

/**
 * Raised when the API returns a 429 Too Many Requests, and retries are exhausted.
 */
export class RateLimitError extends DFError {}

/**
 * Raised when the API returns a 4xx or 5xx status code.
 */
export class APIError extends DFError {
  constructor(
    message: string,
    public readonly statusCode?: number,
    public readonly details?: Record<string, any>
  ) {
    super(message);
  }
}

File: go/exceptions.go
package dataflare


// DFError is the base error for all Dataflare SDK errors.
type DFError struct {
	Message string
}

func (e *DFError) Error() string {
	return e.Message
}

// AuthenticationError is raised when the API key is invalid or expired.
type AuthenticationError struct {
	DFError
}

func NewAuthenticationError(msg string) *AuthenticationError {
	return &AuthenticationError{DFError{Message: msg}}
}

// RateLimitError is raised when the request limit is exceeded.
type RateLimitError struct {
	DFError
}

func NewRateLimitError(msg string) *RateLimitError {
	return &RateLimitError{DFError{Message: msg}}
}

// APIError is raised for general API failures (5xx).
type APIError struct {
	DFError
	StatusCode int
}

func NewAPIError(msg string, code int) *APIError {
	return &APIError{DFError{Message: msg}, code}
}
